### DISCUSSION TITLE
Deprecating `z.discriminatedUnion`. To be replaced by `z.switch`

### DISCUSSION THREAD

**Discussion by JacobWeisenburger** [colinhacks/zod#2109] [General]
Originally posted by @colinhacks
https://github.com/colinhacks/zod/issues/2106

---

I'm deprecating `z.discriminatedUnion` in favor of a "switch" API that's cleaner and more generalizable. You can dynamically "switch" between multiple schemas at parse-time based on the input.

```ts
const schema = z.switch(()=>{
  return Math.random() ? z.string() : z.number()
});
```

I expand more on the `z.switch` API later. Let's talk about `z.discriminatedUnion`.

## Why

 `z.union` naively tries each union element until parsing succeeds. That's slow and bad. Zod needed some solution.

`z.discriminatedUnion` was a mistake. The original `z.discriminatedUnion` API was merged without my approval. I'd entirely missed a long discussion and review process between other contributors and the PR author, and frankly I felt bad about coming in and undoing all that. The API was good but I had reservations about the implementation. It required fiddly recursive logic to exrtract a literal discriminator key from each union element.

<img width="584" alt="Screenshot 2023-02-27 at 2 07 02 AM" src="https://user-images.githubusercontent.com/3084745/221534750-4652a10c-0c6d-4c62-a06b-918ca112c250.png">

It's a bad sign when a method or class requires weird recursive traversal of other schemas. For starters, Zod is designed to be *subclassable*. Users can theoretically subclass `ZodType` to implement custom schema types. But logic like this `instanceof` switch statement don't and can't account for any user-land schema types. 

But the main problem is just that this kind of pattern is bad and introduces a lot of edge cases. It means that only certain kinds of schemas are allowed as discriminators, and others will fail in unexpected ways. There are now dozens of issues that have been opened regarding these various edge cases. The PRs attempting to solve this problem are irredeemably complex and introduce even more edge cases.

- https://github.com/colinhacks/zod/issues/1773
- https://github.com/colinhacks/zod/issues/1667
- https://github.com/colinhacks/zod/issues/1384
- https://github.com/colinhacks/zod/issues/1757
- https://github.com/colinhacks/zod/issues/1504
- https://github.com/colinhacks/zod/issues/1136
- https://github.com/colinhacks/zod/issues/1477
- https://github.com/colinhacks/zod/issues/1424
- https://github.com/colinhacks/zod/pull/1213
- https://github.com/colinhacks/zod/issues/1490
- https://github.com/colinhacks/zod/issues/1181
- https://github.com/colinhacks/zod/issues/1075
- https://github.com/colinhacks/zod/issues/1444
- https://github.com/colinhacks/zod/issues/1868
- https://github.com/colinhacks/zod/issues/1878

> The `.deepPartial` API has this same problem. I'm deprecating it for the same reason.

Many of those issues are asking for non-literal discriminator types:

```ts
type MyUnion = 
  | { type: "a", value: string }
  | { type: "b", value: string }
  | { type: "c", value: string }
  | { type: string, value: string }
  | { type: null, value: string }
  | { type: undefined, value: string }
  | { type: MyEnum, value: string }
  | { type: { nested: string }, value: string }
  | { type: number[], value: string };
```

Imagine each of those elements are represented with Zod schemas. Zod would need to extract the `type` field from each of these elements and find a way to match the incoming `input.type` against those options. In the general case, Zod would extract the `type` field from the shape of each component `ZodObject` and check `input.type` against those schemas until a match is found. At that point, we're back to doing a parse operation for each element of the union, which is what `z.discriminatedUnion` is supposed to avoid doing. (It's still doing less work than the naive `z.union` but still.)

Another issue is composability. The existing API expects the second argument to be an array of `ZodObject` schemas.

```ts
z.discriminatedUnion("type", [
  z.object({ type: z.literal("a"), value: z.string() }),
  z.object({ type: z.literal("b"), value: z.string() }),
])
```

This isn't composable, in that you can't nest discriminated unions or add additional members. 

```ts
const ab = z.discriminatedUnion("type", [
  z.object({ type: z.literal("a"), value: z.string() }),
  z.object({ type: z.literal("b"), value: z.string() }),
]);

const abc = z.discriminatedUnion("type", [
  ab,
  z.object({ type: z.literal("c"), value: z.string() }),
]);
```

- https://github.com/colinhacks/zod/issues/1618
- https://github.com/colinhacks/zod/issues/1610
- https://github.com/colinhacks/zod/issues/1884

Yes, Zod could support both `(ZodObject | ZodDiscriminatedUnion)[]` as union members, but that requires additional messy logic that reflects a more fundamental problem with the API. It also makes increasingly difficult to enforce typesafety on the union - it's important that all union elements have a `type` property, otherwise the union is no longer discriminable.

## Replacement: `z.switch`

```ts
const schema = z.switch(input => {
  return (typeof input) === "string" ? z.string() : z.number();
})

schema.parse("whatever"); // string | number
```

A discriminated union looks like this:

```ts
const schema = z.switch((input) => {
  switch(input.key){
    case "a":
      return z.object({ key: z.literal("a"), value: z.string() })
    case "b":
      return z.object({ key: z.literal("b"), value: z.number() })
    default:
      return z.never()
  }
});
schema.parse({ /* data */ });
// { key: 'a', value: 'asdf' } | { key: 'b', value: number }
```

Ultimately the `z.switch` API is a far more explicit and generalizable API. Zod doesn't do any special handling. The user specifies exactly how the `input` will be used to select the schema. `z.switch()` accepts a function. The `ResultType` of that function is inferred. It will be the union of the schema types returned along all code paths in the function. For instance:

```ts
const schema = z.switch(()=>{
  return Math.random() ? z.string() : z.number()
});
```

Zod sees that the return type of the switcher function is `ZodString | ZodNumber`. The result of the `z.switch` is `ZodSwitch<ZodString | ZodNumber>`. The result of `schema.parse(...)` is `string | number`.

You can represent discriminated unions explicitly like this:

```ts
const schema = z.switch((input) => {
  switch(input.key){
    case "a":
      return z.object({ key: z.literal("a"), value: z.string() })
    case "b":
      return z.object({ key: z.literal("b"), value: z.number() })
    default:
      return z.never()
  }
});
schema.parse({ /* data */ });
// { key: 'a', value: 'asdf' } | { key: 'b', value: number }
```

This can be written in a more condensed form like so:

```ts
const schema = z.switch((input) => ({
  a: z.object({ key: z.literal("a"), value: z.string() }),
  b: z.object({ key: z.literal("b"), value: z.number() }),
}[input.key as string]));
  
schema.parse({ key: 'a', value: 'asdf' });
// { key: 'a', value: 'asdf' } | { key: 'b', value: number }
```

It's marginally more verbose. It's also explicit, closes 30+ issues, eliminates a lot of hairy logic, and lets Zod represent the full scope of TypeScript's type system. `z.discrimininatedUnion` is too fragile and causes too much confusion so it needs to go.

**Comment by suhjohn** [comment]
Are there any plans to support this soon?

**Reply by eshalkowski** [reply]
wondering if I should use `discriminationUnion` now or is `switch` available?

**Reply by erwol** [reply]
An MR was created last January, but it received little feedback. https://github.com/colinhacks/zod/pull/3171
There is also no documentation for `switch`, nor does the above MR include it. `discriminationUnion` should stay for a while as it was not even marked as `@deprecated` yet.

**Comment by tkbeal** [comment]
Is this still being worked on? I love this proposed API and would support the addition of switch()
